resources/agent.py

```diff
@@ -113,38 +117,70 @@ def post(self, id):
             return {'message': 'No file part'}, 400
 
         files = request.files.getlist('file')
+        repo = request.form.get('repo')
 
         file_contents=[]
         for file in files:
-            filename = file.filename
-            if not any([filename.endswith(filetype) for filetype in [".txt", ".pdf", ".md", ".rst"]]):
+            full_path = file.filename
+            file_id = self.get_file_id(repo, full_path)
+            if not any([file_id.endswith(filetype) for filetype in [".txt", ".pdf", ".md", ".rst"]]):
                 return {'message': 'Unsupported file type uploaded'}, 400
 
             file_content = file.stream.read()
 
-            file_contents.append([filename, file_content])
+            file_contents.append([file_id, full_path, file_content])
 
         # Add filenames to the DB
-        new_filenames = agent.filenames.copy()
+        new_full_paths = agent.filenames.copy()
         for fileinfo in file_contents:
-            new_filenames.append(fileinfo[0])
-        agent.filenames = new_filenames
+            new_full_paths.append(fileinfo[0])
+        agent.filenames = new_full_paths
         db.session.commit()
 
         def generate_progress():
-            for filename, file_content in file_contents:
-                yield json.dumps({"file": filename, "step": "start"}) + "\n"
-                extracted_text = text_extractor(filename, file_content)
-                yield json.dumps({"file": filename, "step": "text_extracted"}) + "\n"
+            for _, full_path, file_content in file_contents:
+                yield json.dumps({"file": full_path, "step": "start"}) + "\n"
+                extracted_text = text_extractor(full_path, file_content)
+                yield json.dumps({"file": full_path, "step": "text_extracted"}) + "\n"
 
                 # Only generate embeddings when there is actual texts
                 if len(extracted_text) > 0:
-                    vector_interface.add_document(extracted_text, id, filename)
-                    yield json.dumps({"file": filename, "step": "embedding_created"}) + "\n"
+                    vector_interface.add_document(extracted_text, id, repo, full_path)
+                    yield json.dumps({"file": full_path, "step": "embedding_created"}) + "\n"
 
-                yield json.dumps({"file": filename, "step": "end"}) + "\n"
+                yield json.dumps({"file": full_path, "step": "end"}) + "\n"
 
         return Response(generate_progress(), mimetype='application/json')
+    
+    def delete(self, id):
+        full_path = request.json.get("full_path")
+        repo = request.json.get("repo")
+        metadata = {"repo": repo, "agent_id": id, "full_path": full_path}
+        query = text(f"SELECT id FROM langchain_pg_embedding WHERE cmetadata='{json.dumps(metadata)}';")
```

Comment:
I was looking at the langchain-postgres code and saw its concept of "Collections"... for example, you can init PGVector with a specific collection name:

https://github.com/langchain-ai/langchain-postgres/blob/main/langchain_postgres/vectorstores.py#L348

And you can delete documents only from that collection if you desire:
https://github.com/langchain-ai/langchain-postgres/blob/main/langchain_postgres/vectorstores.py#L629


Would using collections be an alternative to adding `repo` to the metadata? Do you have any thoughts?

Reply:
I have repo there because I was going off some assumptions on how files would get imported. What I was assuming was that if an agent has knowledge from multiple sources (repo's in this case), there could be overlap in the filename and path. i.e. two repos having a README.md in the root wouldn't be distinguishable.
That could be a collection name instead of metadata though

Reply:
Nitpick, but for now let's just call a repo a "source" ... because at the moment it looks like we may pull in docs from all kinds of various "sources" and those may not necessarily be repos... they might come from an s3 bucket for example.